Build and Test Best Practices (#1313)
zowe/community: Technical-Steering-Committee/best-practices/contrinuous-integration.md

# Continuous Integration

## Prerequisites
All Zowe repositories containing source code that requires compilation, packaging, testing or other type of verification, should define a continuos integration process that uses FLOSS automation tools, free to acquire and use for any purposes. It is recommended to choose tools already popular within Zowe, such as Gradle, npm, GitHub Actions, etc. For specific needs, it is possible to select alternatives, given they provide unique and necessary advantages.

## Compilation

> If a Zowe repository requires source code compilation, this repository should include compilers, scripts and tools that automate the process, handle dependencies and produce project artifacts.

Author: Resolved in the latest commit.

## Packaging

> If a Zowe module is intended to be delivered as a deployable installation packed, its corresponding repositories should contain scripts and tools that group the necessary  project artifacts included or produced by compilation into a single or multiple packages targeted for their deployment environments.

Reviewer: ```suggestion
If a Zowe module is intended to be delivered as a deployable installation package, its corresponding repositories should contain scripts and tools that group the necessary  project artifacts included or produced by compilation into a single or multiple packages targeted for their deployment environments.
```

Reviewer: Maybe a shorter version would be sufficient: 
```suggestion
If a Zowe module is intended to be delivered as a deployable installation packed, its corresponding repositories should contain scripts and tools that when executed/used produce automatically these artifacts.
```

Author: Resolved in the latest commit.

## Testing

> Every Zowe module or repository should contain a holistic and automated test suit that is executed on every code change or repository build, i.e. compilation and packaging. Ideally the automated test suit should consist of unit tests to cover the modules being developed or modified, as well as integration tests and system tests to cover interactions with dependent systems. The tests should cover both functional, i.e. validating its functionality, as well as non-functional, i.e. validating performance, reliability, resilience, etc. 

Reviewer: Suggestion: skip "i.e. compilation and packaging". Some repositories may not require compilation and some builds after a change is pushed to a branch may not end with full packaging.

Author: Resolved in the latest commit.

> Every Zowe module or repository should contain a holistic and automated test suit that is executed on every code change or repository build, i.e. compilation and packaging. Ideally the automated test suit should consist of unit tests to cover the modules being developed or modified, as well as integration tests and system tests to cover interactions with dependent systems. The tests should cover both functional, i.e. validating its functionality, as well as non-functional, i.e. validating performance, reliability, resilience, etc. 

Reviewer: dependent systems vs. own dependencies

Author: I don't understand this note. Are you suggesting to change the sentence to: "...as well as integration tests and system tests to cover interactions with dependent systems versus own dependencies." It would not make much sense to me.

> Every Zowe module or repository should contain a holistic and automated test suit that is executed on every code change or repository build, i.e. compilation and packaging. Ideally the automated test suit should consist of unit tests to cover the modules being developed or modified, as well as integration tests and system tests to cover interactions with dependent systems. The tests should cover both functional, i.e. validating its functionality, as well as non-functional, i.e. validating performance, reliability, resilience, etc. 

Reviewer: Maybe performance,reliability etc. tests should not necessarily be executed after every code change.

Author: Resolved in the latest commit.

> Every Zowe module or repository should contain a holistic and automated test suit that is executed on every code change or repository build, i.e. compilation and packaging. Ideally the automated test suit should consist of unit tests to cover the modules being developed or modified, as well as integration tests and system tests to cover interactions with dependent systems. The tests should cover both functional, i.e. validating its functionality, as well as non-functional, i.e. validating performance, reliability, resilience, etc. 

Reviewer: ```suggestion
Every Zowe module or repository should contain a holistic and automated test suit, which should cover both functional, i.e. validating its functionality, as well as non-functional, i.e. validating performance, reliability, resilience, etc. Ideally the automated test suit should consist of unit tests to cover the modules being developed or modified, as well as integration tests and system tests to cover interactions with the module's dependencies. The functional tests should be executed on every code change or repository build and the non functional tests should be executed at least when the build is run to produce relase artifacts, but preferably more ofted to allow early discovery of non-functional issues. 
```

Author: Resolved in the latest commit.

## Code Analysis

> It is required, whenever reasonable, each repository to make use of static code analysis and vulnerability scans per tools availability and requirement. The recommended tools are SonarCloud and CodeQL and should be configured to perform scans for each change. All reported issues should be addressed either by resolving them or by indicating an exception justification.

Reviewer: 1)IMO it is always reasonable to perform code analysis. What exceptions are justifiable to not do so?
2) A minor suggestion: It would more precise to say that the pipelines should be configured to run the corresponding scanning tools instead of the tools themselves to be intercepting the buildin pipelines.

```suggestion
It is required, each repository to make use of static code analysis and vulnerability scans per tools availability and requirement. The recommended tools are SonarCloud and CodeQL and the build pipelines should be configured to run the tools and perform scans for each change. All reported issues should be addressed either by resolving them or by indicating an exception justification.
```

Author: Resolved in the latest commit.

## Quality Gate

> Test coverage for all Zowe code should be no less than 80%. Every change made in a Zowe repository should only be introduced together with a suit of corresponding tests. Every change made in a Zowe repository should require the execution of any and all relevant regression testing in dependent modules. A change cannot be merged if any of the new or existing tests fail.

Reviewer: ```suggestion
Test coverage for all Zowe code should be no less than 80%. Every change made in a Zowe repository should only be introduced together with a suit of corresponding tests. Every change made in a Zowe repository should trigger the execution of any and all relevant regression testing in dependent modules. A change cannot be merged if any of the new or existing tests fail.
```

Author: Resolved in the latest commit.
